src/Backups/BackupImpl.cpp

```diff
@@ -166,6 +170,21 @@ void BackupImpl::open(const ContextPtr & context)
         params.backup_info = *base_backup_info;
         params.open_mode = OpenMode::READ;
         params.context = context;
+        params.use_same_s3_credentials_for_base_backup = use_same_s3_credentials_for_base_backup;
+
+        if (!params.use_same_s3_credentials_for_base_backup)
+        {
+            switch (open_mode)
+            {
+                case OpenMode::WRITE:
+                    writer->substituteCredentials(params.backup_info);
+                    break;
+                case OpenMode::READ:
+                    reader->substituteCredentials(params.backup_info);
```

Comment:
It is not easy to read this "substituteCredentials()" logic.

I think if we add a function `void BackupInfo::copyS3CredentialsTo(BackupInfo & dest) const`  to `BackupInfo`, and also pass this backup's info to the constructor `BackupImpl` instead of `backup_name_for_logging_` then this code can be rewritten simply as

```
if (base_backup_info)
{
    if (params.use_same_s3_credentials_for_base_backup)
        backup_info.copyS3CredentialsTo(*base_backup_info);

    BackupFactory::CreateParams params;
    params.backup_info = *base_backup_info;
    params.open_mode = OpenMode::READ;
    params.context = context;

    base_backup = BackupFactory::instance().createBackup(params);
    ...
```

Reply:
Ok, updated.